Refactor inheritance_chain to inheritance_tree

mm304321141 opened the pull request.

mm304321141 committed:
Add StopWatch for BUILD_VERSION_TIME

mm304321141 committed:
Add VersionBuilderTest.HugeLSM test

mm304321141 committed:
WIP: refactor inheritance_chain to inheritance_tree

levichen94 committed:
WIP: Refactor inheritance_chain to inheritance_tree 2

mm304321141 committed:
Refactor inheritance_chain to inheritance_tree 3

oujiaxin commented on db/compaction.cc:
```diff
@@ -85,6 +85,97 @@ const char* CompactionTypeName(CompactionType type) {
+          }
+        }
+      }
+      if (size + count * 2 != inheritance_tree->size()) {
```
这里count只包括了子树的叶子节点吧，子树的内部节点在哪加的？等式看上去不成立吧

oujiaxin commented on include/rocksdb/table_properties.h:
```diff
@@ -253,8 +254,8 @@ struct TablePropertiesBase {
 
-  // Inheritance chain
-  std::vector<uint64_t> inheritance_chain;
+  // Inheritance tree
```
如何升级？怎么兼容老的Blob？

mm304321141 replied:
老数据是排序的数组，比如
老数据是 [1,2,3]
我们升级转换成
[3,2,1,1,2,3]
即：一颗只有单链的树

Commit:
minor fix of ut

mm304321141 committed:
Fix LazyCompaction old file mark

mm304321141 committed:
Fix BuildInheritanceTree bug

mm304321141 committed:
Add CFOptions::optimize_range_deletion
Add CFOptions::target_blob_file_size
Add CFOptions::max_blob_files
Add CFOptions::max_dependence_blob_overlap
Optimize GC get key

mm304321141 committed:
Refactor VersionBuilder
Add chash

mm304321141 committed:
Refactor VersionBuilder 2

mm304321141 committed:
More details for table_file_creation

mm304321141 committed:
Add SeparationType support for compaction_job

mm304321141 committed:
Fix MemTableList::TryInstallMemtableFlushResults stuck bug

mm304321141 committed:
Add CFOptions::blob_file_defragment_size
Remove CFOptions::ttl
Remove FIFOCompactionStyle

mm304321141 committed:
Refactor CompactionPicker::PickCompositeCompaction for file_number_score

mm304321141 committed:
Fix kCompactionRebuildBlob bug

Commit:
Minor fix

mm304321141 committed:
Refactor TerarkZipTableOptions::terarkZipMinLevel add -2

mm304321141 committed:
Fix huge manual compaction bug

Commit:
fix unittest

mm304321141 committed:
Format Options dump & remove max_blob_files
